### ISSUE TITLE
Some questions about the correctness of the code

### ISSUE THREAD

**Issue by scytdk** [HKUDS/LightGCL#10]
Dear anthors,
        I am interested in the simple while effective approach you propose. In February, I noticed this paper and download its code from https://anonymous.4open.science/r/LightGCL/. Recently I wanna do some improvements based on this work, but I notice that the code of the initial version may be incorrect? For InfoNCE loss, the code implementation is as following:

u_mask = (torch.rand(len(uids))>0.5).float().cuda(self.device)
gnn_u = nn.functional.normalize(self.Z_u_list[l][uids],p=2,dim=1)
hyper_u = nn.functional.normalize(self.G_u_list[l][uids],p=2,dim=1)                
hyper_u = self.Ws[l-1](hyper_u)
pos_score = torch.exp((gnn_u*hyper_u).sum(1)/self.temp)
neg_score = torch.exp(gnn_u @ hyper_u.T/self.temp).sum(1)
loss_s_u = ((-1 * torch.log(pos_score/(neg_score+1e-8) + 1e-8))*u_mask).sum()

The neg_score should be "torch.exp(gnn_u @ self.G_u_list[l].T/self.temp).sum(1)" instead of the code above in my opinion.
And I am pretty confused about it and whether the code is correct. If not correct, how can you get such state-of-art performance?
Look forward to your reply soon.

**Comment by Rick-Cai** [comment]
Hi, thank you for reading our paper. If I understand your question correctly, you are asking why the neg_score is calculated with only part of the users but not all the users? This is because calculating the InfoNCE loss on all the users is very computationally expensive, so in each batch we only sample part of the users to calculate this loss. This is not a matter of correct or incorrect, because in each batch, different users will be sampled randomly, and after many batches, every user will be involved eventually. It's just a batch training strategy to reduce computation in each batch. Please let me know if I misunderstood your question.

**Comment by scytdk** [comment]
Thanks for your timely and patient reply!  I understand the batch strategy, but I does not express my doubts clearly.
I notice that in the github version, the neg_score is gained by:
neg_score = torch.log(torch.exp(G_u_norm[uids] @ E_u_norm.T / self.temp).sum(1) + 1e-8).mean()
which is the dot product of G_u with "uids" and the total E_u with all uids, and it may be different from the initial version released on https://anonymous.4open.science/r/LightGCL/:
neg_score = torch.exp(gnn_u @ hyper_u.T/self.temp).sum(1)
which is the dot product of G_u with "uids" and E_u with "uids". It seems to be a wrong implementation of InfoNCE loss since the denominator part seems to focus on the sampled users instead of the whole users. While I use the whole users to modify it, I notice gradient explosion in training and this is truly disappointing.
So I am wondering if there are some issues with my understanding. Hope for your reply!

**Comment by Rick-Cai** [comment]
Actually, the "denominator part focusing on the sampled users instead of the whole users" is not a "wrong" implementation of InfoNCE loss, but is also a batch strategy. Because as you have noticed, if the denominator takes all users into it, the number will be too large and the gradient will be too large. This "batched InfoNCE" might not exactly match the InfoNCE formula shown in the paper, which implies to include all users, but the effect will be similar after many batches.

As for the difference between this repo and https://anonymous.4open.science/r/LightGCL/, yes, after we submitted to ICLR, we found that our experimental setting is a bit out-dated. Specifically, we were using a sampling strategy that samples pos and neg items according to users, which is different from other researchers' work which directly sample the interactions in the graph. Therefore, we migrated our code to the new setting. In the new setting, the gradients are different, so there are changes in hyperparameter choices, and also some minor adjustment in model details, including the implementation of InfoNCE loss. But these adjustments do not affect the main framework proposed in our paper.
